encodings/dict/src/array.rs

```diff
@@ -97,8 +102,25 @@ impl CanonicalVTable<DictArray> for DictEncoding {
                 let canonical_values: Array = array.values().into_canonical()?.into_array();
                 take(canonical_values, array.codes())?.into_canonical()
             }
-            // Non-string case: take and then canonicalize
-            _ => take(array.values(), array.codes())?.into_canonical(),
+            DType::Primitive(ptype, _)
+                // TODO(alex): handle nullable codes & values
+                if array.values().len() <= 32 && *ptype != PType::F16
+                    && !array.codes().dtype().is_nullable()
+                    && !array.values().dtype().is_nullable() =>
```

Comment:
at this stage you should check `all_valid` since we want to use this codepath if there are no nulls, non nullable is a stronger assertion and we would be missing out on some perf